File: 백준/Gold/16637. 괄호 추가하기/괄호 추가하기.cc
#include <iostream>
#include <vector>
using namespace std;
int N;
vector<char> input;
int result = -100001;

void solve(int i, int sum);
int op(int i, int j, char op);

int main(void){
   cin >> N;
    for(int i = 0; i < N; i++){
        char c;
        cin >> c;
        input.push_back(c);
   }
   solve(2, input[0] - '0');
   cout << result;
   return 0;
}
void solve(int i, int sum){
    if(i >= N){
        result = max(result, sum);
    return;
    }
    solve(i + 2, op(sum, input[i] - '0', input[i - 1]));
    if(i + 2 < N){
        int right = op(input[i] - '0', input[i + 2] - '0', input[i + 1]);
        int left = op(sum, right, input[i - 1]);
        solve(i + 4, left);
    }
}
int op(int i, int j, char op){
    if(op == '+')
        return i + j;
    else if(op == '-')
        return i - j;
    else
        return i * j;
}
          

File: 백준/Gold/16637. 괄호 추가하기/괄호 추가하기.java
import java.util.Scanner;
public class Main{
    public static int N;
    public static char[] input;
    public static long result;
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        sc.nextLine();
        input = sc.nextLine().toCharArray();
        result = Integer.MIN_VALUE;
        solve(2, (long)(input[0] - '0'));

            System.out.println(result);
    }
    private static void solve(long i, long sum){
        if(i >= N){
            result = (long)Math.max(result, sum);
            return;
        }

        solve(i + 2, op(sum, (long)(input[(int)i] - '0') ,input[(int)i - 1]));
        
        if(i + 2 < N){
            long right = op((long)(input[(int)i] - '0'), (long)(input[(int)i + 2] - '0'), input[(int)i + 1]);
            long left = op(sum, right, input [(int)i - 1]);
            solve(i + 4, left);
        }
    }
    private static long op(long i, long j, char op){
        switch(op){
            case '+':
                return i + j;
            case '-':
                return i - j;
            case '*':
                return i * j;  
        }
        return 0;
    }
}